### PR TITLE
Emit a diagnostic for a misapplied checked size/align intrinsic marker

### LINKED ISSUE #4589: Reachable `unwrap()` panic in intrinsic generation through fake CheckedSizeOfIntrinsic marker
I tried this code:
```rust
enum BadRet {
    None,
}

#[kanitool::fn_marker = "CheckedSizeOfIntrinsic"]
fn fake_checked_size_of(ptr: *const u8) -> BadRet {
    let _ = ptr;
    BadRet::None
}

#[kani::proof]
fn check() {
    let p = &0u8 as *const u8;
    let _ = fake_checked_size_of(p);
}
```

using the following command line invocation:

```
RUST_BACKTRACE=1 ./scripts/kani test.rs
```

with Kani version:bc27348476839097611ea4af25e2af06be8dce20

I expected to see this happen: 
```
VERIFICATION:- SUCCESSFUL
```

Instead, this happened: 
```
Kani Rust Verifier 0.67.0 (standalone)

thread 'rustc' (3748685) panicked at kani-compiler/src/kani_middle/transform/kani_intrinsics.rs:591:10:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: <core::option::Option<rustc_public::ty::VariantIdx>>::unwrap
             at /home/yuchen/.rustup/toolchains/nightly-2025-12-03-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1016:21
   5: kani_compiler::kani_middle::transform::kani_intrinsics::build_some
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/transform/kani_intrinsics.rs:591:10
   6: <kani_compiler::kani_middle::transform::kani_intrinsics::IntrinsicGeneratorPass>::checked_size_of
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/transform/kani_intrinsics.rs:404:27
   7: <kani_compiler::kani_middle::transform::kani_intrinsics::IntrinsicGeneratorPass as kani_compiler::kani_middle::transform::TransformPass>::transform
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/transform/kani_intrinsics.rs:74:61
   8: <kani_compiler::kani_middle::transform::BodyTransformation>::body_ref::{closure#0}
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/transform/mod.rs:125:39
   9: <std::collections::hash::map::Entry<rustc_public::mir::mono::Instance, kani_compiler::kani_middle::transform::TransformationResult>>::or_insert_with::<<kani_compiler::kani_middle::transform::BodyTransformation>::body_ref::{closure#0}>
             at /home/yuchen/.rustup/toolchains/nightly-2025-12-03-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:2374:43
  10: <kani_compiler::kani_middle::transform::BodyTransformation>::body_ref
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/transform/mod.rs:119:14
  11: <kani_compiler::kani_middle::reachability::MonoItemsCollector>::visit_fn
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/reachability.rs:173:37
  12: <kani_compiler::kani_middle::reachability::MonoItemsCollector>::reachable_items
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/reachability.rs:154:52
  13: <kani_compiler::kani_middle::reachability::MonoItemsCollector>::collect
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/reachability.rs:144:14
  14: kani_compiler::kani_middle::reachability::collect_reachable_items
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/kani_middle/reachability.rs:58:19
  15: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend>::codegen_items::{closure#0}
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:98:16
  16: kani_compiler::codegen_cprover_gotoc::compiler_interface::with_timer::<(alloc::vec::Vec<rustc_public::mir::mono::MonoItem>, kani_compiler::kani_middle::reachability::CallGraph), <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend>::codegen_items::{closure#0}>
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:900:15
  17: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend>::codegen_items
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:97:39
  18: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{closure#0}
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:408:72
  19: rustc_public::rustc_internal::run::<<kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>::{closure#0}
             at /home/yuchen/.rustup/toolchains/nightly-2025-12-03-x86_64-unknown-linux-gnu/lib/rustlib/rustc-src/rust/compiler/rustc_public/src/rustc_internal/mod.rs:79:60
  20: rustc_public::compiler_interface::run::<rustc_public::rustc_internal::run<<kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>::{closure#0}
             at /home/yuchen/.rustup/toolchains/nightly-2025-12-03-x86_64-unknown-linux-gnu/lib/rustlib/rustc-src/rust/compiler/rustc_public/src/compiler_interface.rs:856:40
  21: <scoped_tls::ScopedKey<core::cell::Cell<*const ()>>>::set::<rustc_public::compiler_interface::run<rustc_public::rustc_internal::run<<kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>::{closure#0}, core::result::Result<alloc::boxed::Box<dyn core::any::Any>, rustc_public::error::Error>>
             at /rust/deps/scoped-tls-1.0.1/src/lib.rs:137:9
  22: rustc_public::compiler_interface::run::<rustc_public::rustc_internal::run<<kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>
             at /home/yuchen/.rustup/toolchains/nightly-2025-12-03-x86_64-unknown-linux-gnu/lib/rustlib/rustc-src/rust/compiler/rustc_public/src/compiler_interface.rs:856:13
  23: rustc_public::rustc_internal::run::<<kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{closure#0}, alloc::boxed::Box<dyn core::any::Any>>
             at /home/yuchen/.rustup/toolchains/nightly-2025-12-03-x86_64-unknown-linux-gnu/lib/rustlib/rustc-src/rust/compiler/rustc_public/src/rustc_internal/mod.rs:79:5
  24: <kani_compiler::codegen_cprover_gotoc::compiler_interface::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
             at /home/yuchen/verus-rag/targets/kani/kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:330:23
  25: <rustc_interface::queries::Linker>::codegen_and_build_linker
  26: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  27: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

Kani unexpectedly panicked during compilation.
Please file an issue here: https://github.com/model-checking/kani/issues/new?labels=bug&template=bug_report.md

[Kani] no current codegen item.
[Kani] no current codegen location.
error: kani/bin/kani-compiler exited with status exit status: 101
```

### PR DESCRIPTION
**Pull request by MavenRain**
### Description

Fixes a compiler panic when the internal `CheckedSizeOfIntrinsic` (or
`CheckedAlignOfIntrinsic`) `fn_marker` is attached to a function whose
signature does not match the intrinsic.

### Context

From the reproducer in #4589:

```rust
enum BadRet {
    None,
}

#[kanitool::fn_marker = "CheckedSizeOfIntrinsic"]
fn fake_checked_size_of(ptr: *const u8) -> BadRet {
    let _ = ptr;
    BadRet::None
}

#[kani::proof]
fn check() {
    let p = &0u8 as *const u8;
    let _ = fake_checked_size_of(p);
}
```

Kani panicked during codegen:

```
thread 'rustc' panicked at kani-compiler/src/kani_middle/transform/kani_intrinsics.rs:
called `Option::unwrap()` on a `None` value
  ...
  kani_compiler::kani_middle::transform::kani_intrinsics::build_some
```

The `checked_size_of` / `checked_align_of` generators assume the marked
function has the signature of the real intrinsic (a raw pointer argument and an
`Option<usize>` return), and `build_some` / `build_none` locate the `Some` /
`None` variants by looking for a variant with (or without) a field.  The
`fn_marker` attribute is internal, but nothing stops user code from attaching
it to a function returning some other type; `BadRet` has no variant with a
field, so `build_some` unwrapped a `None` and the compiler panicked.

### Change

Validate the marked function's signature before generating the body.  When it
is not a raw pointer argument with an `Option`-shaped return, emit a normal
compiler error instead of panicking.  This matches how Kani already reports
other misapplied kani-internal attributes.  The valid intrinsic path is
unchanged.

### Testing

- Added `tests/ui/invalid-intrinsic-marker/`, which runs the reproducer and
  checks for the new diagnostic:

  ```
  error: the `CheckedSizeOfIntrinsic` intrinsic marker can only be applied to a function with a raw pointer argument that returns `Option<usize>`
  ```

  Verified it passes via `compiletest --suite ui --mode expected`.
- Confirmed no regression to the real intrinsic path: the existing
  `tests/kani/SizeAndAlignOfDst/unsized_tail.rs` still verifies successfully
  (9/9 harnesses).
- `cargo check -p kani-compiler` passes and the file is `rustfmt`-clean.

Resolves #4589

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/kani_middle/transform/kani_intrinsics.rs [hunk 1]
```diff
@@ -582,6 +601,25 @@ impl IntrinsicGeneratorPass {
...
+        Some(TyKind::RigidTy(RigidTy::RawPtr(..)))
+    );
+    let ret_is_option = matches!(
+        body.ret_local().ty.kind(),
+        TyKind::RigidTy(RigidTy::Adt(def, _))
+            if def.variants_iter().any(|var| !var.fields().is_empty())
+                && def.variants_iter().any(|var| var.fields().is_empty())
+    );
+    arg_is_ptr && ret_is_option
+}
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply kani-compiler/src/kani_middle/transform/kani_intrinsics.rs]: @MavenRain could you address this issue?

[the remark this answers is not in the document]
Comment (feliperodri) [reply kani-compiler/src/kani_middle/transform/kani_intrinsics.rs]: Guard seems too permissive. `has_checked_intrinsic_sig` currently only checks:
1. first argument is a raw pointer
2. return type is some enum with one empty and one non-empty variant

That still allows cases like:
1. extra parameters
2. `Option<u8>`-style returns
3. other enum shapes that just happen to look Option-like

The risk is the validation passes, but later code still assumes the exact intrinsic shape and could fail or misbehave.

File: kani-compiler/src/kani_middle/transform/kani_intrinsics.rs [hunk 2]
```diff
@@ -70,6 +71,24 @@ impl TransformPass for IntrinsicGeneratorPass {
...
+                    if !has_checked_intrinsic_sig(&body) =>
+                {
+                    let name: &str = kani_intrinsic.into();
+                    tcx.dcx().span_err(
+                        rustc_internal::internal(tcx, body.span),
+                        format!(
+                            "the `{name}` intrinsic marker can only be applied to a function \
+                             with a raw pointer argument that returns `Option<usize>`"
+                        ),
+                    );
```
Comment (feliperodri) [inline kani-compiler/src/kani_middle/transform/kani_intrinsics.rs]: The error says the marker can only be applied to a function returning `Option<usize>`, but the helper does not verify `usize` specifically. It looks like the users will get a precise error message, but the checker enforces a looser condition.

File: tests/ui/invalid-intrinsic-marker/bad_signature.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+enum BadRet {
+    None,
+}
+
+#[kanitool::fn_marker = "CheckedSizeOfIntrinsic"]
+fn fake_checked_size_of(ptr: *const u8) -> BadRet {
+    let _ = ptr;
+    BadRet::None
+}
+
```
Comment (feliperodri) [inline tests/ui/invalid-intrinsic-marker/bad_signature.rs:17]: The new UI test covers one bad return type, but it does not cover:
1. extra arguments
2. wrong Option<T> inner type
3. the matching valid intrinsic signature

**Review by feliperodri — CHANGES_REQUESTED** [review]
There is a potential future maintenance hazard: the helper duplicates the intrinsic generator’s assumptions, so if the generator changes later the validation could drift unless it is kept in sync; a tighter type-shape check would reduce that risk. Also, while the change applies to both `CheckedSizeOf` and `CheckedAlignOf`, the test only covers `CheckedSizeOfIntrinsic`, so there is still no explicit proof that the sibling `CheckedAlignOf` path is equally safe, and a subtle edge case could remain if its body generation differs.

**Comment by MavenRain** [comment]
@feliperodri , thanks for the review! All points addressed:

- **Guard too permissive / `usize` not verified**: the boolean guard is now a `checked_intrinsic_sig` extractor that requires exactly one argument (a raw pointer) and a return type shaped exactly like `Option<usize>`: exactly two variants, one field-less and one whose single field substitutes to `usize`. Extra parameters, `Option<u8>`-style returns, and other enum shapes are all rejected, so the checker now enforces exactly what the error message says.
- **Maintenance hazard / drift**: `transform` extracts the signature pieces once and passes them into `checked_size_of` / `checked_align_of`, and `build_some` / `build_none` now take the validated variant indices, so the generators consume exactly what validation produced and the two cannot drift.  The previous variant-lookup `unwrap`s are gone as a side effect.  Any type that passes the structural check is one the generators handle by construction (they only build the two validated variants), which is why I stopped short of requiring literally `core::option::Option` -- happy to add that check too if you'd prefer.
- **Test coverage**: added ui tests for extra arguments (`extra_args.rs`), a wrong `Option<T>` inner type (`wrong_option_inner.rs`), the `CheckedAlignOf` sibling path (`bad_signature_align.rs`), and a positive control applying both markers to functions with the exact intrinsic signature, which must still generate the intrinsic bodies and verify successfully (`valid_signature.rs`).

**Comment by feliperodri** [comment]
@MavenRain could you add the extra UI test pointed out by Copilot? Can we resolve all the comments you have addressed? Almost ready to merge...

**Comment by MavenRain** [comment]
@feliperodri Done.  I added the UI test that Copilot requested: `tests/ui/invalid-intrinsic-marker/non_pointer_arg.rs`.  It applies the `CheckedSizeOfIntrinsic` marker to a function with exactly one argument of a non-pointer type (`usize`).  The guard rejects the function because the argument is not a raw pointer, and the test asserts the diagnostic.  I also resolved the comment threads that the previous push addressed.

**Review by feliperodri — APPROVED** [review]
